[SDCP-969] - Add to Planning not pulling slugline automatically

eos87 opened the pull request:
### Purpose
Resolves race condition where field values arrive after component initialization, causing "Add to Planning" to show empty slugline (and other fields) instead of auto-populating from the news item.

### What has changed
- Adds `getDerivedStateFromProps` to sync late-arriving prop values while preserving user input once they start typing.

### To reproduce the issue
- Go to SD CP
- Go to Monitoring menu 
- Click the three dot and then “Add to Planning” and then click + to add a new planning item
- Add to Planning modal dialog does not populate the `slugline`

Resolves: [SDCP-969]


[SDCP-969]: https://sofab.atlassian.net/browse/SDCP-969?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eos87 committed:
Add to Planning not pulling slugline automatically

Resolves race condition where field values arrive after component initialization, causing "Add to Planning" to show empty slugline instead of auto-populating from the news item.

Solution adds getDerivedStateFromProps to sync late-arriving prop values
while preserving user input once they start typing.

SDCP-969

eos87 committed:
Add `immer` dependency and Update lockfile to version 3

SDCP-969

Revert "Add `immer` dependency and Update lockfile to version 3"

This reverts commit d7be50f24a979365d64fa911a0d71e5e6cfb5b34.

Add immer dependency and update packages

eos87 committed:
Reset user modification state on item change

Added componentDidUpdate to reset userHasModified to false when the item prop changes, ensuring late-arriving prop values can repopulate the field. This prevents stale user modification state when switching items.

SDCP-969

eos87 committed:
Merge tag 'v3.0.0-beta.2' into hg/SDCP-969-add-to-planning-slugline-missing-1

thecalcc commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "@sourcefabric/common": "0.0.65",
     "@sourcefabric/draft-js": "^0.11.5",
     "dompurify": "^1.0.11",
+    "immer": "^10.1.1",
```
why do we need immer here?

eos87 replied:
it's breaking the build and I saw it was included in release/3.1 but I couldn't cherry pick it (it's used from superdesk-client-core)

thecalcc commented on client/components/fields/editor/base/text.tsx:
```diff
@@ -58,8 +81,16 @@ export class EditorFieldText extends React.Component<IEditorFieldTextProps, ISta
+    componentDidUpdate = (prevProps: Readonly<IEditorFieldTextProps>): void => {
+        // Make sure to reset user modification state when item changes
+        // so that late-arriving prop values can populate the field again
+        if (!isEqual(prevProps.item, this.props.item)) {
```
i'm just wondering isn't there a better way to see if we have changes - the `this.props.diff` if we have access to it, should hold the changes

eos87 replied:
I'm not sure but I think the `diff` would be different when you change an input so it's not what I'm looking for here. My intention is to reset the `userHasModified` when the item being edited/created is different from the previous so the data arriving late from props can update the state in `getDerivedStateFromProps`

thecalcc reviewed (commented):
small comments

eos87 committed:
Try to fix lint error on GH actions

eos87 committed:
Revert "Try to fix lint error on GH actions"

This reverts commit d731c31ac533411dd897e16abc9ea45e0653b8ba.

petrjasek reviewed (commented):
I wonder if some parent component shouldn't rather wait until it has all data it needs and only render those components after

eos87 commented:
> I wonder if some parent component shouldn't rather wait until it has all data it needs and only render those components after

Ideally yes and I considered it but, it is not simple. The problem is that one of the parent components intentionally does multiple state updates, therefore achieving that would be very tricky without refactoring many pieces. 

This is what I'm referring to:
https://github.com/superdesk/superdesk-planning/blob/154db9922837f824112435b1cea6f34d5155d5e6/client/components/Main/ItemEditor/Editor.tsx#L107-L158

eos87 committed:
Testing with npm ci

eos87 commented:
This PR depends on https://github.com/superdesk/superdesk-planning/pull/2567

eos87 committed:
Merge remote-tracking branch 'origin/release/3.0' into hg/SDCP-969-add-to-planning-slugline-missing-1

eos87 committed:
Update install steps for Node 14 and 22 in CI workflow

Adjusted the CI workflow to use 'npm ci' for Node 14 and 'npm install' for Node 22 to ensure compatibility with different Node versions during client setup.

petrjasek commented on client/components/fields/editor/base/text.tsx:
```diff
@@ -58,8 +81,16 @@ export class EditorFieldText extends React.Component<IEditorFieldTextProps, ISta
+        // Make sure to reset user modification state when item changes
+        // so that late-arriving prop values can populate the field again
+        if (!isEqual(prevProps.item, this.props.item)) {
+            this.setState({userHasModified: false});
```
I wonder now why not set it in the `getDerivedStateFromProps` ?

eos87 replied:
because `getDerivedStateFromProps` is static and do not have access to current props or state.

eos87 replied:
but thinking it twice, I think I could solve it with `getDerivedStateFromProps`. Let me give it a try.

eos87 replied:
@petrjasek nope, my idea did not work. It prevents the user from typing.

petrjasek commented on client/components/fields/editor/base/text.tsx:
```diff
@@ -58,8 +81,16 @@ export class EditorFieldText extends React.Component<IEditorFieldTextProps, ISta
+    componentDidUpdate = (prevProps: Readonly<IEditorFieldTextProps>): void => {
+        // Make sure to reset user modification state when item changes
+        // so that late-arriving prop values can populate the field again
+        if (!isEqual(prevProps.item, this.props.item)) {
```
would be good not to use `isEqual` here but check using `===` or compare item id if that's when it should reset the flag

eos87 replied:
I didn't use `===` because it does not perform deep equality in dict like objects. As for the `_id`, from what I saw, the `_id` is not always provided/reliable. Although the fact that was provided or not could be sufficient in this case. Do you think that's the case?

petrjasek replied:
it seems like overkill to use deep equality on item if it's editing single field, and as that editor could be there multiple times and item structure with associations can be quite deep I would avoid it. if it's about switching when item changes imo `===` should be enough, we probably don't have 2 different items which would pass the `===` test. using `_id` would be ok too in that case probably

eos87 replied:
makes sense. I've made the change and went with `_id`

eos87 committed:
Optimize item change detection in EditorFieldText

Replaced deep equality check with direct _id comparison in componentDidUpdate to improve performance.

SDCP-969

eos87 committed:
Remove unused isEqual import from text editor

SDCP-969
